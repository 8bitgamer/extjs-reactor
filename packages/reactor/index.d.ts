import React = require('react')

export interface InstallOptions {
    /**
     * Adds a stylesheet that mimics an Ext JS Viewport by setting the html, body, and react root element to height: 100%.
     * Set this to true when using an Ext JS component at the root of your app.
     */
    viewport: boolean;
}

/**
 * Configures React.
 */
export function install(options: Partial<InstallOptions>): void;

/**
 * Creates a react component for a given Ext JS component.
 * For example: const Grid = reactify('grid');
 * @param target xtype or instance of Ext.Class.
 */
export function reactify<Class, Props, State>(target: string | Function): Class & React.Component<Props, State>;
export function reactify<Class1, Props1, State1, Class2, Props2, State2>(target1: string | Function, target2: string | Function): [Class1 & React.Component<Props1, State1>, Class2 & React.Component<Props2, State2>];
export function reactify<Class1, Props1, State1, Class2, Props2, State2, Class3, Props3, State3>(target1: string | Function, target2: string | Function, target3: string | Function): [Class1 & React.Component<Props1, State1>, Class2 & React.Component<Props2, State2>, Class3 & React.Component<Props3, State3>];
export function reactify<Class1, Props1, State1, Class2, Props2, State2, Class3, Props3, State3, Class4, Props4, State4>(target1: string | Function, target2: string | Function, target3: string | Function, target4: string | Function): [Class1 & React.Component<Props1, State1>, Class2 & React.Component<Props2, State2>, Class3 & React.Component<Props3, State3>, Class4 & React.Component<Props4, State4>];
export function reactify(...targets: (string | Function)[]): React.Component<any, any>[];
