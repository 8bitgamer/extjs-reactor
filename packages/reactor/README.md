# Ext JS Reactor

The @extjs/reactor package is a custom renderer that makes it easy to use [Ext JS](https://www.sencha.com/products/extjs) components in your [React](https://facebook.github.io/react) app. Once installed, all JSX tags starting with "x-" are resolved to Ext JS components by xtype.

## Requirements

* Ext JS 6.2+
* Sencha Cmd 6.2+
* React 15.3.2+ (peer dependency)

## Installation

```bash
# Be sure to install react>=15.3.2 before
npm install --save-dev @extjs/reactor
```

## Getting Started

We recommend you start by cloning the [boilerplate project](https://github.com/sencha/extjs-reactor/tree/master/packages/reactor-boilerplate) and following the instructions there.

## Basic Concepts

First, install the custom renderer.  We recommend doing this in your index.js file (your webpack entry point).  This only needs to be done once in your app.

```jsx
import install from '@extjs/reactor';
install();
```

All JSX tags that start with "x-" will be mapped to Ext JS components by xtype. You can override the default prefix when calling `install`, for example: `install("ext-")`.  Here's a minimal React app that renders an Ext.Panel:

```jsx
import React from 'react';
import ReactDOM from 'react-dom';
import install from '@extjs/reactor';

// Install the Ext JS custom renderer
install();

// When Ext JS loads, initialize our React app
Ext.onReady(() => {
    ReactDOM.render(
        (
            <x-panel title="React Ext JS">
                Hello World!
            </x-panel>
        ),
        document.getElementById('root')
    );
});
```


React props are converted to Ext JS configs.  Here's a typical use of an Ext.grid.Grid:

```jsx
import React, { Component } from 'react';

export default class MyComponent extends Component {
    render() {        
        return (
            <x-grid
                plugins={[                
                    { type: 'columnresizing' }
                ]}
                columns={[
                    { text: 'Name', dataIndex: 'name' },
                    { text: 'Email', dataIndex: 'email' }
                ]}
                store={{
                    fields: ['name', 'email'],
                    data: [
                        { name: 'Tim Smith', email: 'tim101@gmail.com' },
                        { name: 'Jill Lindsey', email: 'jlindsey890@gmail.com' }
                    ]
                }}
            />
        )
    }
}
```


Any prop starting with "on" followed by a capital letter is automatically converted to an Ext JS event listener.  Since Ext JS events are all lower-case, case is not preserved.  You're free to use camel-case, which is common in React.

```jsx
import React, { Component } from 'react';

export default function MyComponent() {
    return (
        <x-slider
            minValue={0}
            maxValue={100}
            onChange={(slider, value) => console.log(`Value set to ${value}`)}
        />
    )
}
```


You can also use a listeners object as is common in traditional Ext JS:

```jsx
import React, { Component } from 'react';

export default function MyComponent() {
    return (
        <x-slider
            minValue={0}
            maxValue={100}
            listeners={{
                change: (slider, value) => console.log(`Value set to ${value}`)
            }}
        />
    )
}
```


Refs point to Ext JS component instances:

```jsx
import React, { Component } from 'react';

export default class MyComponent {
    render() {
        return (
            <x-slider
                ref="slider"
                minValue={0}
                maxValue={100}
                onChange={() => this.onChange()}
            />         
        )
    }

    onChange() {
        console.log('Slider value', this.refs.slider.getValue()); // this.refs.slider is an Ext.slider.Slider
    }
}
```


Select your toolkit, theme, and packages using [@extjs/reactor-webpack-plugin]. The plugin scans your code and only includes the classes you need in the final bundle.  Here's an example:

```JavaScript
const ExtJSReactWebpackPlugin = require('@extjs/reactor-webpack-plugin');

module.exports = {
    ...
    plugins: [
        new ExtJSReactWebpackPlugin({
            sdk: 'ext', // location of Ext JS SDK
            theme: 'theme-material',
            packages: ['charts'],
            output: path.join('build', 'ext')
        })
    ]
    ...
}
```
