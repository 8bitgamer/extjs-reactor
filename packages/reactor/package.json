{
  "name": "@extjs/reactor",
  "version": "0.2.16",
  "description": "Use Ext JS components in React.",
  "main": "dist/index.js",
  "module": "src/index.js",
  "typings": "index.d.ts",
  "scripts": {
    "watch": "babel ./src --out-dir ./dist --source-maps --watch",
    "build": "babel ./src --out-dir ./dist --source-maps",
    "prepublish": "npm run build",
    "create-typings": "node script/create-typings.js",
    "test": "jest"
  },
  "jest": {
    "rootDir": "src",
    "setupFiles": [
      "../jest/environment.js"
    ]
  },
  "keywords": [
    "Sencha",
    "Ext JS",
    "React",
    "ExtReact"
  ],
  "author": "Sencha",
  "license": "MIT",
  "homepage": "https://github.com/sencha/extjs-reactor#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/sencha/extjs-reactor.git"
  },
  "bin": {
    "ext-react": "./bin/ext-react.js"
  },
  "peerDependencies": {
    "react": "^15.4.0",
    "react-dom": "^15.4.0"
  },
  "dependencies": {
    "lodash.capitalize": "^4.2.1",
    "lodash.clonedeepwith": "^4.5.0",
    "lodash.defaults": "^4.2.0",
    "lodash.isequalwith": "^4.4.0",
    "lodash.union": "^4.6.0",
    "minimist": "^1.2.0",
    "mkdirp": "^0.5.1"
  },
  "devDependencies": {
    "@extjs/sencha-cmd": "^6.5.0-177",
    "@types/react": "^0.14.52",
    "@types/react-dom": "^0.14.19",
    "babel-cli": "^6.14.0",
    "babel-jest": "^18.0.0",
    "babel-preset-es2015": "^6.14.0",
    "babel-preset-react": "^6.11.1",
    "babel-preset-stage-2": "^6.13.0",
    "jest": "^15.1.1",
    "react": "^15.4.0",
    "react-addons-test-utils": "^15.4.0",
    "react-dom": "^15.4.0",
    "react-test-renderer": "^15.3.2",
    "sinon": "^1.17.6",
    "typescript": "^2.2.0-dev.20161202"
  }
}
