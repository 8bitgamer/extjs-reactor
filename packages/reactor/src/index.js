import ExtJSComponent from './ExtJSComponent';
import ReactHostComponent from 'react/lib/ReactHostComponent';

const original = ReactHostComponent.createInternalComponent;

/**
 * Configures React to resolve jsx tags.
 * @param {String} [prefix="x-"] All jsx tags with this prefix will be resolved to Ext JS xtypes
 * @param {String} [viewport=true] Adds a stylesheet that mimics an Ext JS Viewport
 *    by setting the html, body, and react root element to height: 100%. Set this to true when using an
 *    Ext JS component at the root of your app.
 */
export default function({ prefix="x-", viewport=false } = {}) {
    ReactHostComponent.createInternalComponent = function(element) {
        if (element.type.startsWith(prefix)) {
            return new ExtJSComponent(element);
        } else {
            return original(element);
        }
    };

    let css = '.react-extjs-host { height: 100%; }';

    if (viewport) {
        css += '\nhtml, body, div[data-reactroot] { height: 100%; }';
    }

    const style = document.createElement('style')
    style.innerHTML = css;
    document.head.appendChild(style);
}
