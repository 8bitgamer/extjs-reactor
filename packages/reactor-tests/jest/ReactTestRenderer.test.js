import React from 'react';
import { create } from 'react-test-renderer'
import { reactify } from '@extjs/reactor';

const Container = reactify('Container');
const Button = reactify('Button');
const [ Grid, Column, Toolbar ] = reactify('Grid', 'Column', 'Toolbar');

describe('ReactTestRenderer snapshots', () => {
    
    it('should support html child elements', () => {
        const result = create(
            <Container>
                <div className="foo" style={{color: 'red'}}>
                    <b>F</b>oo
                </div>
                <div className="bar">Bar</div>
            </Container>
        );

        expect(result).toMatchSnapshot();
    });

    it('should support function props', () => {
        const result = create(
            <Button text="Button" handler={() => console.log('clicked')}/>
        )

        expect(result).toMatchSnapshot();
    });

    it('should omit complex props', () => {
        const store = new Ext.data.Store({
            fields: ['name']
        });

        class Foo {

        }

        const result = create(
            <Grid 
                title="Employees"
                store={store}
                foo = {new Foo()}
            />
        )

        expect(result).toMatchSnapshot();
    });

    it('should support nested ExtReact components', () => {
        const store = new Ext.data.Store({
            fields: ['name']
        });

        const result = create(
            <Grid 
                title="Employees"
                store={store}
            >
                <Toolbar>
                    <Button text="Add Record"/>
                </Toolbar>
                <Column dataIndex="name" text="Name"/>
            </Grid>
        )

        expect(result).toMatchSnapshot();
    });

    it('should support ExtReact components inside html elements', () => {
        const result = create(
            <div>
                <Container>
                    <Button text="Button"/>
                    <div>Test</div>
                </Container>
            </div>
        );

        expect(result).toMatchSnapshot();
    })
});