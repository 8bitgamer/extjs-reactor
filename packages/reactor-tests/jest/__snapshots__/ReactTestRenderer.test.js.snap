// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactTestRenderer snapshots should omit complex props 1`] = `
<Grid
  foo={
    Object {
      "$className": "Foo",
    }
  }
  store={
    Object {
      "$className": "Ext.data.Store",
    }
  }
  title="Employees"
/>
`;

exports[`ReactTestRenderer snapshots should support ExtReact components inside html elements 1`] = `
<div>
  <Container>
    <Button
      text="Button"
    />
    <div>
      Test
    </div>
  </Container>
</div>
`;

exports[`ReactTestRenderer snapshots should support function props 1`] = `
<Button
  handler={[Function]}
  text="Button"
/>
`;

exports[`ReactTestRenderer snapshots should support html child elements 1`] = `
<Container>
  <div
    className="foo"
    style={
      Object {
        "color": "red",
      }
    }
  >
    <b>
      F
    </b>
    oo
  </div>
  <div
    className="bar"
  >
    Bar
  </div>
</Container>
`;

exports[`ReactTestRenderer snapshots should support nested ExtReact components 1`] = `
<Grid
  store={
    Object {
      "$className": "Ext.data.Store",
    }
  }
  title="Employees"
>
  <Toolbar>
    <Button
      text="Add Record"
    />
  </Toolbar>
  <Column
    dataIndex="name"
    text="Name"
  />
</Grid>
`;
